src/wallet/wallet.cpp

```diff
@@ -1906,7 +1906,12 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         // Check if any destination contains a witness program:
         int witnessversion = 0;
         std::vector<unsigned char> witnessprogram;
-        if (recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) {
+        const bool is_wit = recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram);
+        if (is_wit && witnessversion >= 1 && GetScriptPubKeyMan(OutputType::BECH32M, true)) {
+            // If this is a segwit v1+ address and this wallet supports bech32m addresses, use bech32m
+            return OutputType::BECH32M;
+        } else if (is_wit) {
+            // Otherwise if this is witness, use bech32
```

Comment:
This assumes there is always a bech32 descriptor, which may not be the case if you're using a custom taproot descriptor without fallback.

```patch
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 3eca506f08..c0b28acd47 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1911,9 +1911,11 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         if (is_wit && witnessversion >= 1 && GetScriptPubKeyMan(OutputType::BECH32M, true)) {
             // If this is a segwit v1+ address and this wallet supports bech32m addresses, use bech32m
             return OutputType::BECH32M;
-        } else if (is_wit) {
-            // Otherwise if this is witness, use bech32
+        } else if (is_wit && GetScriptPubKeyMan(OutputType::BECH32, true)) {
+            // Otherwise if this is witness, use bech32 if we have a descriptor:
             return OutputType::BECH32;
+        } else {
+            return m_default_address_type;
         }
     }
```

Reply:
The original code did not check for a bech32 `ScriptPubKeyMan`, but I have added that check. I did not include the `else` suggested in your diff as that would be incorrect.